pkg/controller/association/secret.go

```diff
@@ -0,0 +1,181 @@
+		Version: ver,
+		URL:     info.URL,
+		// points the auth secret to the custom secret
+		AuthSecretName: assocRef.SecretName,
+		AuthSecretKey:  authPasswordUnmanagedSecretKey,
+		CACertProvided: info.CaCert != "",
+	}
+	// points the ca secret to the custom secret if needed
+	if expectedAssocConf.CACertProvided {
+		expectedAssocConf.CASecretName = assocRef.SecretName
+	}
+
+	return expectedAssocConf, err
+}
+
+func GetAuthFromUnmanagedSecretOr(client k8s.Client, unmanagedAssocRef commonv1.ObjectSelector, other func() (string, string, error)) (string, string, error) {
+	if unmanagedAssocRef.IsObjectTypeSecret() {
+		info, err := GetUnmanagedAssociationConnectionInfoFromSecret(client, unmanagedAssocRef)
+		if err != nil {
+			return "", "", err
+		}
+		return info.Username, info.Password, nil
+	}
+	return other()
+}
+
+// UnmanagedAssociationConnectionInfo holds connection information stored in a custom Secret to reach over HTTP an Elastic resource not managed by ECK
+// referenced in an Association. The resource can thus be external to the local Kubernetes cluster.
+type UnmanagedAssociationConnectionInfo struct {
+	URL      string
+	Username string
+	Password string
+	CaCert   string
+}
+
+// GetUnmanagedAssociationConnectionInfoFromSecret returns the UnmanagedAssociationConnectionInfo corresponding to the Secret referenced in the ObjectSelector o.
+func GetUnmanagedAssociationConnectionInfoFromSecret(c k8s.Client, o commonv1.ObjectSelector) (*UnmanagedAssociationConnectionInfo, error) {
+	var secretRef corev1.Secret
+	secretRefKey := o.NamespacedName()
+	if err := c.Get(context.Background(), secretRefKey, &secretRef); err != nil {
+		return nil, err
+	}
+	url, ok := secretRef.Data["url"]
+	if !ok {
+		return nil, fmt.Errorf("url secret key doesn't exist in secret %s", o.SecretName)
+	}
+	username, ok := secretRef.Data["username"]
+	if !ok {
+		return nil, fmt.Errorf("username secret key doesn't exist in secret %s", o.SecretName)
+	}
+	password, ok := secretRef.Data[authPasswordUnmanagedSecretKey]
+	if !ok {
+		return nil, fmt.Errorf("password secret key doesn't exist in secret %s", o.SecretName)
+	}
+
+	ref := UnmanagedAssociationConnectionInfo{URL: string(url), Username: string(username), Password: string(password)}
+	caCert, ok := secretRef.Data[certificates.CAFileName]
+	if ok {
+		ref.CaCert = string(caCert)
+	}
+
+	return &ref, nil
+}
+
+func (r UnmanagedAssociationConnectionInfo) Request(dialer net.Dialer, path string, jsonPath string) (string, error) {
+	req, err := http.NewRequest("GET", r.URL+path, nil) //nolint:noctx
+	if err != nil {
+		return "", err
+	}
+
+	req.SetBasicAuth(r.Username, r.Password)
+
+	var caCerts []*x509.Certificate
+	if r.CaCert != "" {
+		caCerts, err = certificates.ParsePEMCerts([]byte(r.CaCert))
+		if err != nil {
+			return "", err
+		}
+	}
+
+	httpClient := common.HTTPClient(dialer, caCerts, client.DefaultESClientTimeout)
+
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		return "", err
```

Comment:
Errors that happen here are hard to diagnose because they only show up in the logs. I think we might want to surface this better in the status or at least an event. Also my impression was that it does not retry after a few attempts and just stays in `Failed` status but I have to verify that once more.

Reply:
Good points, adressed in 3da52ff & 9780417.

```
82s  Warning  AssociationError   kibana/test  Association backend for elasticsearch is not configured
39s  Warning  AssociationError   kibana/test  Failed to reconcile external resource "es-monitor-ref": error requesting /, statusCode = 401
40s  Warning  AssociationError   kibana/test  Failed to reconcile external resource "es-monitor-ref": invalid character '<' looking for beginning of value
1s   Normal   AssociationStatusChange   kibana/test  Association status changed from [Failed] to [Established]
```

Not sure about the wording. It would be longer but more correct to say:
Failed to reconcile external resource referenced in secret "es-monitor-ref"?

Reply:
Sure, 90b0c58 puts the full URL.